frontend/javascripts/oxalis/model/helpers/nml_helpers.js

```diff
@@ -372,22 +372,24 @@ function isTreeConnected(tree: Tree): boolean {
 
   if (tree.nodes.size() > 0) {
     // Get the first element from the nodes map
-    const nodeQueue = [Number(tree.nodes.keys().next().value)];
+    const nodesToBeVisited = new Set([Number(tree.nodes.keys().next().value)]);
     // Breadth-First search that marks all reachable nodes as visited
-    while (nodeQueue.length !== 0) {
-      const nodeId = nodeQueue.shift();
+    while (nodesToBeVisited.size > 0) {
+      // $FlowFixMe nodeId will be a number as we checked the size of the set first
+      const nodeId: number = nodesToBeVisited.values().next().value;
```

Comment:
I'm not sure I completely understand the problematic scenario. We are adding visited nodes to the `visitedNodes` set to ensure that we don't see the same node twice, **but** we are setting the visited flag not until the node is actually processed. So, that means that we can add a node potentially twice to the queue, right? This is avoided with the new nodesToBeVisited queue which guarantees uniqueness?

Couldn't we just change the semantics of `visitedNodes` to something like `seenNodes` for which we do `seenNodes.set(nodeId, true);` when adding a node to the nodeQueue? That way, we would also guarantee that a node is only ever added once to the queue, and we don't have two data structures which handle uniqueness?

Reply:
In essence we would only move/copy line 380 to line 387 and 389?

Reply:
Yes this should work as well and is the better solution :)